File: quiz-nodejs-engine/src/models/Question.ts
import 'reflect-metadata';
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  OneToMany,
} from 'typeorm';
import { QuizResponse } from './QuizResponse';

@Entity('questions')
export class Question {
  @PrimaryGeneratedColumn()
  id!: number;

  @Column({ type: 'varchar', length: 2000 })
  questionText!: string;

  @Column({ type: 'varchar', length: 500 })
  optionA!: string;

  @Column({ type: 'varchar', length: 500 })
  optionB!: string;

  @Column({ type: 'varchar', length: 500 })
  optionC!: string;

  @Column({ type: 'varchar', length: 500 })
  optionD!: string;

  @Column({ type: 'varchar', length: 500, nullable: true })
  optionE?: string;

  @Column({ type: 'varchar', length: 1 })
  correctAnswer!: string;

  @Column({ type: 'varchar', length: 2000, nullable: true })
  explanation?: string;

  @Column({ type: 'varchar', length: 100, nullable: true })
  section?: string;

  @Column({ type: 'varchar', length: 50, nullable: true })
  difficulty?: string;

  @Column({ type: 'varchar', length: 255, nullable: true })
  sourceFile?: string;

  @Column({ type: 'integer', default: 1 })
  usageCycle!: number;

  @Column({ type: 'integer', default: 0 })
  timesUsed!: number;

  @Column({ type: 'datetime', nullable: true })
  lastUsedAt?: Date;

  @CreateDateColumn()
  createdAt!: Date;

  @OneToMany(() => QuizResponse, (response) => response.question, { lazy: true })
  responses!: Promise<QuizResponse[]>;
}

File: quiz-engine/quiz-engine-springboot/src/main/java/com/quizengine/entity/Question.java
package com.quizengine.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "questions", indexes = {
    @Index(name = "idx_section", columnList = "section"),
    @Index(name = "idx_difficulty", columnList = "difficulty"),
    @Index(name = "idx_usage_cycle", columnList = "usage_cycle")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Question {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String questionText;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String optionA;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String optionB;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String optionC;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String optionD;

    @Column(columnDefinition = "TEXT")
    private String optionE;

    @Column(nullable = false)
    private String correctAnswer;

    @Column(columnDefinition = "TEXT")
    private String explanation;

    @Column
    private String section;

    @Column
    private String difficulty;

    @Column
    private String sourceFile;

    @Column(name = "usage_cycle", nullable = false)
    @Builder.Default
    private Integer usageCycle = 1;

    @Column(nullable = false)
    @Builder.Default
    private Integer timesUsed = 0;

    @Column
    private LocalDateTime lastUsedAt;

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @OneToMany(mappedBy = "question", cascade = CascadeType.REMOVE)
    private List<QuizResponse> responses;
}
